Review pass (adversarial + testing + maintainability) done; fixes pushed as `6694a8ba`. The adversarial pass found a **security hole I introduced**, and it was right to block.

## `--env` could exfiltrate host secrets

`EnvValue` is an untagged enum whose variants deserialize through `with_expand_envs`, so a value is re-interpreted **every time the descriptor is deserialized** — in the coordinator, then again in the daemon. I verified this against `dora-message` directly rather than taking the report on faith:

```
wire "${DORA_AUTH_TOKEN}"  ->  s3cr3t-daemon-token     ← expands in the RECEIVING process
wire "p$ssw0rd"            ->  ERROR: no matching variant
wire "1.10"                ->  1.1                     ← silent coercion
```

So `dora start flow.yml --env 'X=${DORA_AUTH_TOKEN}'` would have handed a node the **coordinator's or daemon's** token. `strip_denied_env` cannot stop it — that guard matches key *names*, and the key here is innocuous. This was new with the flag: YAML `env:` values are expanded CLI-side at descriptor read, so they reach the wire already substituted; `--env` was the only way to put an unexpanded `$` on it.

Escaping doesn't fix it (two deserialization hops — a `$$` surviving one is expanded on the next), so `--env` now refuses any value that doesn't survive a serialize/deserialize round-trip verbatim. One check covers all three shapes above, and the error names the key and points at the node `env:` block. Confirmed end to end:

```
$ dora start flow.yml --env 'LEAK=${DORA_AUTH_TOKEN}'
[ERROR] --env `LEAK=${DORA_AUTH_TOKEN}` cannot be represented in the dataflow descriptor…
  hint: a literal `$` cannot be carried through `--env` … Set this one in the node's `env:` block instead.
```

I chose refusal over silent delivery of a different value: rejecting `VERSION=1.10` is annoying, but handing a node `1.1` when the operator typed `1.10` is worse. Four new unit tests pin all of it, including the branch where expansion *succeeds* (the real exfiltration shape, using `PATH` as a stand-in for the secret).

## Other findings applied

- **`dora run --env` had zero coverage** — the entire `descriptor_override` branch was untested. Added an e2e, mutation-verified by passing the session descriptor instead of the merged one (fails with `got "from-yaml"`). It runs local-only, so it can't contend for port 6013.
- **`dora run` validated `--env` after a full build** — a typo cost a compile. Now parses first, like `start`.
- **Marker flush race** in the `start` e2e: the poll checks existence, but Python file objects aren't line-buffered, so it could read a created-but-unflushed file. Now write-then-`os.replace`.
- **Merge-placement premise** now pinned by a unit test asserting `--env` changes the descriptor fingerprint — the ordering constraint in `start_dataflow` was previously guarded only by a comment.
- Flag help documents the wire restrictions plus persistence in the coordinator store / visibility in `ps`.
- Maintainability: idiomatic `Descriptor::blocking_read`, proper imports, comment placement restored around the tokio spawn block, private `merge_env` no longer linked as public API, unix-only rationale on both e2es.

## Flagged, not fixed

`DORA_NODE_CONFIG`, `DORA_ZENOH_*` and `DORA_RUNTIME_CONFIG` are set **before** the user-env loop in the spawner and are absent from `ENV_DENYLIST`, so dataflow-level env can override them — `--env DORA_NODE_CONFIG=…` hands a node a forged config. Pre-existing and already reachable from YAML `env:` today, so not an escalation this PR creates, but the denylist exists as defense-in-depth for spawn integrity. Worth hardening separately rather than inside a CLI-flag PR — happy to file it.

Sweep: `dora-cli` 282 unit tests, `ws-cli-e2e` 27/27, fmt/clippy clean.

--- a/binaries/cli/src/env_overrides.rs
+++ b/binaries/cli/src/env_overrides.rs
@@ -69,7 +69,33 @@ pub fn parse_env_overrides(flags: &[String]) -> Result<BTreeMap<String, EnvValue
 /// `$$` that survives one hop is expanded on the next. Rejecting is the
 /// honest option — silently handing a node a different value than the
 /// operator typed is worse than refusing to start.
+///
+/// The `$` rule is deliberately **syntactic**, not "does this value
+/// survive a round-trip in the CLI's process". That round-trip is not a
+/// sound test of the security property: expansion happens against
+/// *whichever* environment decodes the descriptor, so a CLI-local
+/// variable that refers to itself (`DORA_AUTH_TOKEN='${DORA_AUTH_TOKEN}'`)
+/// makes the CLI-side round-trip an identity — passing the check — while
+/// the wire still carries `${DORA_AUTH_TOKEN}` for the daemon to expand
+/// against the *real* secret. Refusing every `$` closes that fixed-point
+/// bypass by never letting an expandable value reach the wire at all.
 fn ensure_wire_safe(key: &str, value: &str) -> Result<()> {
+    if value.contains('$') {
+        bail!(
+            "--env `{key}={value}` contains `$`, which cannot be carried in a \
+             dataflow descriptor: the process that receives it (coordinator, \
+             then daemon) expands the value against ITS OWN environment, so \
+             `$VAR` would resolve to that host's variable rather than \
+             yours.\n\n  \
+             hint: expand it in your shell first (`--env {key}=\"$VAR\"` \
+             without quotes around the `$`), or set this one in the node's \
+             `env:` block in the dataflow YAML."
+        );
+    }
+
+    // Beyond `$`, the untagged enum coerces numeric-looking strings —
+    // `1.10` decodes as Float(1.1). A round-trip IS a sound test for
+    // that, since no environment lookup is involved.
     let encoded = serde_json::to_string(&EnvValue::String(value.to_string()))
         .with_context(|| format!("failed to encode --env `{key}`"))?;
     match serde_json::from_str::<EnvValue>(&encoded) {
@@ -77,17 +103,15 @@ fn ensure_wire_safe(key: &str, value: &str) -> Result<()> {
         Ok(decoded) => bail!(
             "--env `{key}={value}` would not survive the dataflow descriptor's \
              encoding: nodes would receive `{decoded}` instead.\n\n  \
-             hint: values containing `$` are expanded by the process that \
-             receives the descriptor, and numeric-looking values are coerced \
-             (`1.10` -> `1.1`). Set this one in the node's `env:` block in the \
-             dataflow YAML instead."
+             hint: numeric-looking values are coerced (`1.10` -> `1.1`, \
+             `01234` -> `1234`). Set this one in the node's `env:` block in \
+             the dataflow YAML instead."
         ),
         Err(err) => bail!(
             "--env `{key}={value}` cannot be represented in the dataflow \
              descriptor: {err}.\n\n  \
-             hint: a literal `$` cannot be carried through `--env` — the \
-             descriptor expands it on receipt. Set this one in the node's \
-             `env:` block in the dataflow YAML instead."
+             hint: set this one in the node's `env:` block in the dataflow \
+             YAML instead."
         ),
     }
 }
@@ -169,24 +193,62 @@ mod tests {
     /// guard matches key names, and the key here is innocuous.
     #[test]
     fn rejects_values_that_would_be_env_expanded_on_the_wire() {
-        // Unset variable: expansion fails outright on decode.
-        let result = parse_env_overrides(&flags(&["LEAK=${DORA_TEST_2919_UNSET}"]));
-        let err = result.expect_err("a wire-expanded value must be rejected");
-        assert!(
-            err.to_string().contains("LEAK"),
-            "the error must name the offending key: {err}"
-        );
+        for value in [
+            "LEAK=${DORA_TEST_2919_UNSET}", // unset here, maybe set on the host
+            "LEAK=${PATH}",                 // set here AND on the host
+            "LEAK=$PATH",                   // brace-less form
+            "LEAK=prefix-${PATH}-suffix",   // embedded, not the whole value
+        ] {
+            let result = parse_env_overrides(&flags(&[value]));
+            let err = result.expect_err("a wire-expandable value must be rejected");
+            assert!(
+                err.to_string().contains("LEAK"),
+                "the error must name the offending key: {err}"
+            );
+        }
+    }
 
-        // Set variable: expansion SUCCEEDS and silently substitutes — the
-        // real exfiltration shape, where the deserializing daemon holds
-        // the secret. Rejected because the decoded value differs from
-        // what the operator typed. `PATH` stands in for the secret; it is
-        // set in every environment this runs in.
-        let result = parse_env_overrides(&flags(&["LEAK=${PATH}"]));
+    /// The fixed-point bypass this check exists to close: expansion runs
+    /// against whichever environment decodes the descriptor, so a
+    /// CLI-local variable that refers to itself makes a CLI-side
+    /// round-trip an identity — while the wire still carries `$VAR` for
+    /// the DAEMON to expand against the real secret. A round-trip test
+    /// passes this; a syntactic `$` rule does not.
+    ///
+    /// Uses a scoped child process rather than `set_var`, which is
+    /// `unsafe` in edition 2024 and would race sibling tests.
+    #[test]
+    fn self_referential_local_variable_cannot_smuggle_a_dollar_through() {
+        let exe = std::env::current_exe().expect("test binary path");
+        let output = std::process::Command::new(exe)
+            .args(["--exact", "env_overrides::tests::self_referential_child"])
+            .env("DORA_TEST_2919_SELFREF", "${DORA_TEST_2919_SELFREF}")
+            .env("RUST_TEST_NOCAPTURE", "0")
+            .output()
+            .expect("failed to re-run test binary");
+        assert!(
+            output.status.success(),
+            "the self-referential value was NOT rejected — a `$` reached the \
+             wire.\nstdout:\n{}\nstderr:\n{}",
+            String::from_utf8_lossy(&output.stdout),
+            String::from_utf8_lossy(&output.stderr),
+        );
+    }
+
+    /// Child of `self_referential_local_variable_cannot_smuggle_a_dollar_through`;
+    /// only meaningful with `DORA_TEST_2919_SELFREF` set to its own
+    /// reference, so it is a no-op when run directly.
+    #[test]
+    fn self_referential_child() {
+        let Ok(value) = std::env::var("DORA_TEST_2919_SELFREF") else {
+            return; // not the child invocation
+        };
+        assert_eq!(value, "${DORA_TEST_2919_SELFREF}", "child env not set up");
+        let result = parse_env_overrides(&flags(&["LEAK=${DORA_TEST_2919_SELFREF}"]));
         assert!(
             result.is_err(),
-            "a value that expands to the receiving process's env must be \
-             rejected, got {result:?}"
+            "a self-referential expansion round-trips to itself, so it must be \
+             caught syntactically rather than by comparison, got {result:?}"
         );
     }
 

--- a/binaries/coordinator/src/lib.rs
+++ b/binaries/coordinator/src/lib.rs
@@ -1585,6 +1585,17 @@ async fn start_inner(
 
                                         // Resolve the Node into a ResolvedNode via a
                                         // temporary single-node descriptor.
+                                        //
+                                        // Carry the running dataflow's own `env:`
+                                        // into that descriptor so the added node
+                                        // inherits it exactly as its statically
+                                        // declared peers did — `resolve_aliases_and_set_defaults`
+                                        // performs the dataflow-into-node merge
+                                        // (node keys still win). Passing `None`
+                                        // here meant a node added later silently
+                                        // saw none of the dataflow's environment,
+                                        // including anything set via `dora start
+                                        // --env` (dora-rs/dora#2919).
                                         let tmp_desc = dora_message::descriptor::Descriptor {
                                             nodes: vec![node],
                                             communication: Default::default(),
@@ -1593,7 +1604,7 @@ async fn start_inner(
                                             health_check_interval: None,
                                             strict_types: None,
                                             type_rules: Vec::new(),
-                                            env: None,
+                                            env: dataflow.descriptor.env.clone(),
                                         };
                                         match tmp_desc.resolve_aliases_and_set_defaults() {
                                             Ok(mut resolved_map) => {

--- a/tests/ws-cli-e2e.rs
+++ b/tests/ws-cli-e2e.rs
@@ -1285,6 +1285,154 @@ mod real_dataflow {
         );
     }
 
+    /// #2919 P2: a node added to a RUNNING dataflow must inherit that
+    /// dataflow's env, including anything set via `dora start --env`.
+    ///
+    /// `dora node add` resolves the incoming node through a temporary
+    /// single-node descriptor in the coordinator; that descriptor was
+    /// built with `env: None`, so the dataflow-into-node merge had
+    /// nothing to merge and an added node silently saw none of the
+    /// dataflow's environment. Reproduced as `<unset>` before the fix.
+    ///
+    /// Unix-only for the same reasons as the sibling env tests.
+    #[test]
+    #[cfg(unix)]
+    fn e2e_dynamically_added_node_inherits_dataflow_env() {
+        ensure_built();
+        let dora = dora_bin();
+        start_cluster(&dora);
+        let _guard = ClusterGuard(&dora);
+
+        let target = Path::new(env!("CARGO_MANIFEST_DIR")).join("target");
+        if let Ok(entries) = std::fs::read_dir(&target) {
+            for entry in entries.flatten() {
+                if entry
+                    .file_name()
+                    .to_string_lossy()
+                    .starts_with("dora-addenv-2919-")
+                {
+                    let _ = std::fs::remove_file(entry.path());
+                }
+            }
+        }
+        let stem = format!("dora-addenv-2919-{}", std::process::id());
+        let probe = target.join(format!("{stem}-probe.py"));
+        let base_marker = target.join(format!("{stem}-base.txt"));
+        let added_marker = target.join(format!("{stem}-added.txt"));
+        let yaml = target.join(format!("{stem}.yml"));
+        let added_yaml = target.join(format!("{stem}-added.yml"));
+        std::fs::write(
+            &probe,
+            concat!(
+                "import os, time\n",
+                "out = os.environ['PROBE_OUT']\n",
+                "with open(out + '.tmp', 'w') as f:\n",
+                "    f.write(os.environ.get('E2919_VAR', '<unset>') + '\\n')\n",
+                "    f.write(os.environ.get('E2919_NODE', '<unset>') + '\\n')\n",
+                "os.replace(out + '.tmp', out)\n",
+                "time.sleep(120)\n",
+            ),
+        )
+        .expect("failed to write probe");
+        std::fs::write(
+            &yaml,
+            format!(
+                "nodes:\n  \
+                 - id: base\n    \
+                   path: {probe}\n    \
+                   env:\n      \
+                     PROBE_OUT: {base_marker}\n    \
+                   outputs:\n      - value\n",
+                probe = probe.display(),
+                base_marker = base_marker.display(),
+            ),
+        )
+        .expect("failed to write dataflow yaml");
+        // The added node sets its own E2919_NODE, so this also pins that
+        // node-level env still wins for dynamically added nodes.
+        std::fs::write(
+            &added_yaml,
+            format!(
+                "id: added\n\
+                 path: {probe}\n\
+                 env:\n  \
+                   PROBE_OUT: {added_marker}\n  \
+                   E2919_NODE: node-wins\n\
+                 outputs:\n  - value\n",
+                probe = probe.display(),
+                added_marker = added_marker.display(),
+            ),
+        )
+        .expect("failed to write added node yaml");
+
+        let status = Command::new(&dora)
+            .args([
+                "start",
+                yaml.to_str().unwrap(),
+                "--detach",
+                "--name",
+                "addenv",
+                "--env",
+                "E2919_VAR=from-cli",
+                "--env",
+                "E2919_NODE=cli-loses",
+            ])
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .status()
+            .expect("failed to run dora start");
+        assert!(status.success(), "dora start --env failed");
+
+        // Wait for the base node so the dataflow is up before adding.
+        let deadline = std::time::Instant::now() + Duration::from_secs(30);
+        while !base_marker.exists() {
+            assert!(
+                std::time::Instant::now() < deadline,
+                "base node never spawned (is `python3` installed?)"
+            );
+            std::thread::sleep(Duration::from_millis(200));
+        }
+
+        let output = Command::new(&dora)
+            .args([
+                "node",
+                "add",
+                "--dataflow",
+                "addenv",
+                "--from-yaml",
+                added_yaml.to_str().unwrap(),
+            ])
+            .output()
+            .expect("failed to run dora node add");
+        assert!(
+            output.status.success(),
+            "dora node add failed: {}",
+            String::from_utf8_lossy(&output.stderr)
+        );
+
+        let deadline = std::time::Instant::now() + Duration::from_secs(30);
+        while !added_marker.exists() {
+            assert!(
+                std::time::Instant::now() < deadline,
+                "added node never spawned"
+            );
+            std::thread::sleep(Duration::from_millis(200));
+        }
+        let content = std::fs::read_to_string(&added_marker).expect("failed to read marker");
+        let lines: Vec<&str> = content.lines().collect();
+        assert_eq!(
+            lines.first().copied(),
+            Some("from-cli"),
+            "a node added to a running dataflow must inherit its env \
+             (including `--env`); got {content:?}"
+        );
+        assert_eq!(
+            lines.get(1).copied(),
+            Some("node-wins"),
+            "node-level env: must still win for added nodes; got {content:?}"
+        );
+    }
+
     /// The `dora run` half of #2919, which reaches the daemon by a
     /// different route: `run` builds a `descriptor_override` and hands it
     /// to the in-process `Daemon::run_dataflow`, rather than serializing

